rfcs/SchemaCoordinates.md

```diff
@@ -0,0 +1,372 @@
+   requested**. This may be implemented by incrementing a counter by the name of
+   the schema coordinate for each field executed in a request.
+
+   _Existing implementations: Yelp (internal), Facebook (internal)_
+
+1. GraphiQL and other playgrounds / documentation sites want to show a list of
+   **search results** when a user searches for a type or field name. We can
+   display a list of schema coordinates that match the search term. A schema
+   coordinate can also be used in the hyperlink to form a permalink for
+   documentation for a particular field.
+
+   _Existing implementations: GraphiQL, Apollo Studio (see "Prior Art")_
+
+1. A developer may want to perform **analytics** on all known
+   [persisted queries][apq] - e.g. what are the most commonly used fields across
+   all documents. Schema coordinates may be used as the index/lookup keys when
+   storing this information in the database.
+
+   _Existing implementations: Yelp (internal)_
+
+   [apq]: https://www.apollographql.com/docs/apollo-server/performance/apq/
+
+1. A **GitHub bot** may want to warn developers in a Pull Request comment
+   whenever the schema diff contains a breaking change. Schema coordinates can be
+   used to provide a list of which fields were broken.
+
+   _Existing implementations: GraphQL Inspector (see "Prior Art")_
+
+1. **GraphQL IDEs** (e.g. GraphiQL, GraphQL Playground, Apollo Studio) may wish
+   to display the schema definition type of a node in a query when hovering over
+   it.
+
+   <details>
+   <summary>Example</summary>
+   ![](https://i.fluffy.cc/g78sJCjCJ0MsbNPhvgPXP46Kh9knBCKF.png)
+   </details>
+
+   Schema coordinates can be used to form the left hand side of this popover.
+
+_Existing implementations: Apollo Studio (see "Prior Art")_
+
+## 🚫 What this RFC does _not_ propose
+
+- This does not cover "selectors" or "wildcard" syntax - e.g. `User.*`. _(See
+  alternatives considered.)_
+- There are **no proposed GraphQL language/syntax changes**
+- There are **no proposed GraphQL runtime changes**
+- [Schema coordinate non-goals](#-syntax-non-goals)
+
+## ✨ Worked Examples
+
+For example, consider the following schema:
+
+```graphql
+type Person {
+  name: String
+}
+
+type Business {
+  name: String
+  owner: Person
+}
+
+type Query {
+  searchBusinesses(name: String): [Business]
+}
+```
+
+We can write the following list of Schema Coordinates:
+
+- `Person` uniquely identifies the the "Person" type
+- `Business` uniquely identifies the the "Business" type
+- `Person.name` uniquely identifies the "name" field on the "Person" type
+- `Business.name` uniquely identifies the "name" field on the "Business"
+  type
+- `Business.owner` uniquely identifies the "owner" field on the "Business" type
+- `Query.searchBusinesses` uniquely identifies the "searchBusinesses" field on
+  the "Query" type
+- `Query.searchBusinesses(name:)` uniquely identifies the "name" argument on the
```

Comment:
I'm still not keen on this syntax for arguments. `Query.searchBusinesses.name` feels a bit more natural to me, and is less likely to be mistaken for an actual query fragment or something. But I don't have a compelling argument here, it's mostly just personal preference.

Reply:
If we stick with this syntax we can extend it in future to reference a position in a GraphQL document; e.g.

```
Query.businesses:searchBusinesses(name:).owner:personByOwnerId.email
```

might refer to:

```graphql
{
  businesses: searchBusinesses(name: "Automotive") {
    id
    name
    owner: personByOwnerId {
      id
      name
      email # <<< HERE
    }
  }
}
```

Reply:
Actually that's a poor example (probably you'd just say `businesses.owner.email`); but you could write something like the above expression and have it automatically expand to a similar GraphQL document, similar to how you can use CSS shortcuts to expand to DOM structure using emmet or similar.

Reply:
@eapache Happy to go either way! I'm pretty 50/50 on `Foo.bar.baz` vs `Foo.bar(baz:)`. How do we decide here? Toss a coin? Dance-off battle?

@benjie  On the topic of something like `Query.businesses:searchBusinesses(name:).owner:personByOwnerId.email`, just to clarify:
1) Schema coordinates are supposed to "uniquely identify a type". Given our use cases, it's ideal if there's no normalization that has to be done to get to the "simplest" schema coordinate for the same thing.
2) Tracing the path to a member of a query/document is kinda already a thing with [field paths](https://spec.graphql.org/June2018/#example-90475) (`"path": [ "hero", "heroFriends", 1, "name" ]`).

Perhaps it's possible one day we unify this syntax and expand schema coordinates to support nesting, but I'll gingerly suggest that this is out of scope for this particular RFC? (Worth bearing in mind for a future extension of this syntax tho, along with selecting multiple arguments)

Thoughts?

Reply:
> Dance-off battle?

I'm in 😂 

Seriously though, I think it's probably something worth chatting about at the IRL working group meetings. There might be an obvious consensus there, and if not then we'll probably want to do a slightly more formal pros/cons writeup to justify making a final call (basically the process we're following for input unions, but not nearly so heavy).

> businesses.owner.email

Yeah, the schema-coordinates vs query-path vs result-path distinction is still a little weird... they're definitely distinct but they *feel* like they ought to be related somehow.

Case in point, in the Shopify example I shared above we're technically sharing a weird combination of coordinate and path. For use of deprecated fields we share the plain coordinate `Type.field` but for use of deprecated inputs we append the full input path (so not `InputType.field` but actually `Type.field.argument.inputFieldPath...`.

Reply:
ack! Ok leaving this open for now. I'll proceed with the `Foo.bar(baz:)` style just so we have something written down, but we'll make sure to talk about it next WG meeting. Happy to ctrl-f search and replace it all to `Foo.bar.baz` if that's where we land - should be an easy to change to make now that we're prepared for it.

I'll add this as an open question to the RFC.

Thanks!

Reply:
> Perhaps it's possible one day we unify this syntax and expand schema coordinates to support nesting, but I'll gingerly suggest that this be out of scope for this particular RFC? (Worth bearing in mind for a future extension of this syntax tho, along with selecting multiple arguments)

100%; I was never suggesting adding this to this spec, more that we should make sure that the spec enables future innovation including using it for things _other than_ schema coordinates. To my mind the `(foo:)` syntax is more flexibile in this regard. For example, I can imagine referring to:

1. `Foo.bar(baz:.qux)`: the `qux` field of the input object referred to from the `baz` argument of the `bar` field on the `Foo` type.
2. `Foo.bar(baz:).qux`: the `qux` field on the return type of the `bar` field (with `baz:` argument) of the `Foo` type.
3. `Foo.bar.baz.qux`: the `qux` field of the return type of the `baz` field on the return type of the `bar` field on type `Foo`.

If we were to only use periods then all of these would come out the same as `Foo.bar.baz.qux`, and this ambiguity precludes this kind of reusal of the schema-coordinates syntax for this use case (which is outside the scope of the schema coordinates spec, for sure, but is still a potential use-case for the syntax).

Ultimately what I'm saying is I feel it would be a shame if the syntax could not be re-used outside of schema coordinates.

Reply:
> `Foo.bar(baz:).qux`

I would rewrite this all-periods as `Foo.bar.quz`. I don't see why the argument is relevant if the thing you're trying to reference is the `qux` field. Assuming a well-formed query then you don't even need the argument to uniquely identify the path.

Reply:
Sure, if we're using them as coordinates rather than doing Emmet-style auto-completion or some other use case someone has come up with. Even as simply coordinates though, 1 and 3 are still distinct yet would have the same representation with periods?

Imagine you had the following schema

```graphql
type Foo {
  bar(baz: BazInput): Bar
}

type Bar {
  baz(qux: Int): Baz
}

type Baz {
  qux: Int
}

input BazInput {
  qux: Int
}

type Query {
  foo: Foo
}
```

With periods you might write `Foo.bar.baz.qux`, and that could refer to (using argument syntax) `Foo.bar(baz:.qux)` (input field), `Foo.bar.baz(qux:)` (argument) or `Foo.bar.baz.qux` (field) - three very different things. GraphQL query language looks like the JSON it returns, it makes sense to me that schema coordinates should look like the GraphQL entities they represent.

Reply:
I've added a couple notes in the RFC about this syntax:

1. ```Should arguments look like `Foo.bar.baz` instead of `Foo.bar(baz:)`?``` (we'll discuss in a WG meeting)
2. As an RFC goal: `Schema coordinate syntax is open for extension in the future.` To @benjie (and other's) points, whilst we're not currently supporting fancy extended syntax as described, we could all imagine extending schema coordinate syntax to support multiple arguments, wildcards etc. So if we can do something now even just to give us more flexibility and wiggle room in the future, then that's probably a good thing.

Reply:
Following yesterday's WG (and the thumbs up above) I believe that @eapache is happy to go ahead with the parenthesized syntax.

Additional note; I wrote `Foo.bar(baz:.qux)` above; but I think `Foo.bar(baz.qux:)` has better aesthetics.

Reply:
Bit of a tangent (although it seems like using parens is resolved so it may be safe) - 

In the example of `Foo.bar(baz:).qux`, is the contents of the parens carrying useful information for the coordinates?  

Assuming arguments require parens, would `Foo.bar(baz:).qux` point to the same place as `Foo.bar.qux`?

Reply:
Is it too late to add my opinion? I agree with @eapache the C struct.member notation is much better for several reasons. 
- 1. Exactly that, it is C style syntax so it comes naturally for many devs who work with C style languages such as php. 
- 2. A period is easier and quicker to type than (:). 
- 3. It is 3 non alpha numeric characters instead of 1, which makes a 3 byte difference AND requires a parser while the struct syntax could be implemented in C as a struct without parsing. 
 
This also makes a big difference in embedded and IOT systems which can benefit immensely from graph data structures if they are implemented efficiently. This could literally keep me from using graphQL in a project.

Lastly why call it coordinates? It doesn't have anything to do with position or am I missing something. This constant reinvention of terms in web frameworks has always been a major issue for newcomers. Things should be named for clarification and not because they sound fancy. It is not descriptive of it's purpose. It is used to identify just as it was described in the OP. So why not call it a Field Identifier? Field Descriptor? It may not sound hip or new but I know what it means.

If you go up to non GraphQL literate person you should be able to explain a query without having to explain what a "field coordinate" is after using the word. If you were to ask me to identify the coordinate in those queries I would think you are trolling me and I have been a web dev for 15+ years. I just asked my GF what the Descriptor was and she said "the name after the period". She is a English teacher and doesn't know what a div tag is.

Reply:
I've had some time to chew this over - given discussions last WG meeting, I'm personally pretty settled on `Foo.bar(baz:)`. That's the option I'm proposing in the RFC.

I've captured this discussion in the RFC here: https://github.com/magicmark/graphql-spec/blob/add_field_coordinates/rfcs/SchemaCoordinates.md#field-arguments

Please let me know if anyone feels I missed anything, or if there's more points to add (or existing points to contest!)

Thanks everyone who has helped so far in thinking through this :)

@excitedbox to answer some stuff:

1. On the topic of argument syntax, it is a bit of a tradeoff! I've tried to capture the reasoning in a [new section in the RFC](https://github.com/magicmark/graphql-spec/blob/add_field_coordinates/rfcs/SchemaCoordinates.md#field-arguments). I welcome your followup feedback!
2. For naming, I'll point again to the RFC, as there's [a whole section on this](https://github.com/magicmark/graphql-spec/blob/add_field_coordinates/rfcs/SchemaCoordinates.md#naming). I will say that in the last WG meeting, I literally bought up the point "schema coordinates sounds a bit fancy", but I think (1) ultimately it's a more accurate name than the alternatives I had, and (2) it's reusing prior art (graphql java already used the "coordinates" naming). If you have an alternate suggestion, I'd love to hear your thoughts.

Reply:
(I'm going to leave this conversation thread open till at least the next WG to make sure this discussion remains visible for folks)